# 16주차 BDA
- 지난시간에 이은 t분포 등의 코드 이어하기

In [1]:
import numpy as np
from scipy import stats
import matplotlib.pyplot as plt

## 문제1: 소표본의 t검정(모분산을 모르는 경우)
- BDA의 새로운 학습법으로 벌점에 영향을 미친다고 주장한다.
- 25명의 학회원에게 새로운 학습법을 적용해서 성적을 보니 성적이 70, 표준편차가 10
- 기존의 성적 평균점수가 65
- 이 결과가 유의미한가? 유의수준 0.05에서 설정하여 검정해라!

In [2]:
np.random.seed(111)
sample_mean = 70
sample_std = 10
n = 25

# 귀무가설: 새로운 학습법은 성적에 영향을 끼치지 않는다. 65점이다. 그 이상이 되지 않는다
# 대립가설: 새로운 학습법은 성적에 영향을 준다. 65점 이상이다

data = np.random.normal(loc = sample_mean, scale = sample_std, size = n)
data

array([58.66161666, 73.84319193, 84.96553776, 66.44617703, 62.12466459,
       65.40561092, 69.40831229, 66.45826499, 62.64476945, 58.16060114,
       72.38894134, 64.10079738, 55.59414882, 77.73703108, 59.7203267 ,
       69.0901375 , 74.92003003, 74.24672201, 82.83048815, 73.15986454,
       65.91917798, 69.32052407, 60.47573342, 68.89322598, 75.70594293])

In [3]:
# 모평균 검정을 위한 t검정
t_stats, p_value = stats.ttest_1samp(data, popmean = 65)

print("t통계:",t_stats)
# pvalue : 0.03 = 3%기 때문에 95%신뢰수준일 경우 5%보다 작기때문에 대립가설이 차용된다
print("pvalue:",p_value)

t통계: 2.295521358365354
pvalue: 0.03074411024414607


## t통계와 pvalue

- t검정 결과 -> t통계가 2.29~가 나옴
- p value: 0.03 _ 대립가설 채택가능여부
- 두가지를 통해서 해석

### t통계

- t통계량 표본 평균이 귀무가설과 얼마나 차이가 나는지는 '표준오차단위'로 나타난다
  - 표본 평균이 귀무가설의 평균과 2.295 표준오차의 차이가 있다 (t통계값)

### pvalue
- p-value : 귀무가설이 참이라는 가정하에, 표본 데이터가 관찰된 t통계량보다 극단적인 값을 가질 확률
  - 즉 pvalue가 5%로 보통 되는 이유 = 표본 데이터가 t통계량보다 극단적일 확률이 5%라는 뜻. 95% 확률의 정상적임을 원하는 경우인 것


### 정리하면..!
- 유의수준을 0.05로 유의수준알파(a) 설정.5% 오차 허용
- pvalue 유의수준 비교
  - p-값 0.0307 < 유의수준(0.05)가 작다
  - 유의수준보다 작기 때문에 귀무가설을 기각할 수 있다.
  - 새로운 학습법이 성적향상에 도움을 준다는게 통계적으로 유의미하게 지지한다 (p값이 매우 작아서 아웃라이어이지 않을 확률이 더 크기 때문!)
-------
- 유의 수준 a는 귀무가설을 기각할 때 잘못된 결정을 내릴 확률에( = 1종 오류), 대한 허용 오차다.
  - 보통 5%

- p값 _ 귀무가설이 참이라는 가정하에 우리가 관찰한 데이터가 나타날 확률
- p값이 작다는건 ?  _ 관찰한 데이터, 즉 25명의 학회원 데이터가 귀무가설 하에서는 발생하기가 어려운 데이터임을 의미함

- 왜 유의수준 p값을 비교하나요?
  - 유의수준a, 이를 기준으로 p값을 비교하여 귀무가설을 기각할지에 대한 여부를 결정한다.
  - p값이 유의수준보다 작다(p<a):
    - 관찰된 데이터가 귀무가설 하에서는 발생하기가 매우 어렵다
    -  따라서 대립가설을 채택한다.
    - 즉 대립가설의 발생확률이 높다는 뜻이 된다

  - p값이 유의수준보다 크다(p>=a)
    - 관찰된 데이터 귀무가설 하에도 충분히 발생할 수 있는 범위에 있다는 의미.
    - 대립가설을 기각하는 수치를 의미한다.

### 동전 던지기 예제
1. 귀무가설 :동전던지기는 공정하다 (H0)
2. 대립가설: 동전던지기는 공정하지 않다(H1)

- 10번을 던져 8번의 앞면이 나왔다
- 유의수준 @는 0.05
- p값은 공정한 동전 던지기라는 가정하에, 10번 중에 8번이 앞면이 나오는 확률 계산
- p값이 0.03이라는 값이 나왔다. 공정한 동전던지기라는 가정하에서 10번 중에 8번이 앞면이 나올 확률은 3%
- p값이 유의수준보다 작다 -> 동전 던지기가 공정하다는 귀무가설을 기각하고 동전 던지기가 공정하지 않을 수 있따는 결론이 나온다

#### 결론
1. 유의수준은 우리가 설정한 @값이 작으면 귀무가설을 기각 -> 귀무가설이 참이라는 가정하에 데이터가 발생할 확률이 매우낮다
2. 귀무가설을 기각하고 대립가설을 받아들일 수 있는 근거

## 문제2. 대표본 Z-검정(모분산을 알고있는 경우)
- BDA제품을 만들었는데 생산된 제품의 평균 무게가 50g
- 무작위로 100개의 샘플을 샘플링한 결과, 평균 무게가 52g, 표준편차는 2g
- 모분산이42g이라고 알려져있을 때, 이 결과가 50g과 유의미하게 다른지 검정하라

In [4]:
np.random.seed(111)
sample_mean = 52
sample_std = 2
n = 100

data = np.random.normal(loc = sample_mean, scale = sample_std , size = n)
data

array([49.73232333, 52.76863839, 54.99310755, 51.28923541, 50.42493292,
       51.08112218, 51.88166246, 51.291653  , 50.52895389, 49.63212023,
       52.47778827, 50.82015948, 49.11882976, 53.54740622, 49.94406534,
       51.8180275 , 52.98400601, 52.8493444 , 54.56609763, 52.63197291,
       51.1838356 , 51.86410481, 50.09514668, 51.7786452 , 53.14118859,
       53.83084034, 48.66131899, 52.96542729, 51.37905466, 56.78937946,
       55.10186184, 50.70706965, 50.14212553, 48.69004771, 52.70038501,
       51.71648697, 53.04216357, 51.95819827, 48.5123121 , 50.40168225,
       49.39285914, 52.35621   , 51.33119527, 51.38794574, 51.33518885,
       55.92589468, 53.43848498, 54.28577302, 56.16575394, 49.43070449,
       53.07625522, 51.91092118, 57.19432874, 51.88346744, 50.10942601,
       53.08234478, 51.88998223, 54.24004162, 51.6167141 , 50.77972316,
       51.1108423 , 47.59198141, 51.13865903, 51.14981308, 52.29458399,
       52.84984854, 52.69451549, 51.44752891, 49.59540875, 54.06

- 유의수준 하에 p값이 0.0000001

In [5]:
# 모평균 검정을 위한 z검정 (모분산이 알려져있다)
z_stat = (sample_mean -50)/(sample_std / np.sqrt(n))
p_value = 2*(1-stats.norm.cdf(z_stat))

print("z통계:",z_stat)
print("pvalue:",p_value)

z통계: 10.0
pvalue: 0.0


## 문제3. 모분산에 대한 Chi-square 검정 (소표본, 모분산을 모르는 경우)

#### 이후 보충 코드 올라올 예정.과제에 적용하기

- BDA에서 학생들의 수업 지각 변동성을 줄이기 위해 새로운 프로그램을 개발하여 사용한다.
- 10번 실험을 통해서 시간을 확인해보니 평균 30초 + 표준편차가 2초 정도 나타남.
- 기존의 반응시간이 4.2초로 알려져있을 때 새로운 프로그램의 개발이 학회원들의 수업시간 지각 변동성을 줄이는데 유의미한지 검정하자
- 유의수준 a = 0.05

- H0 = 새로운 프로그램을 통해 학생들의 지각은 달라지지 않았다
- H1 = 새로운 프로그램을 통해 학생들의 지각이 개선되었다(양측)


( 추후 공지로 코드 제공 예정)

In [6]:
# 카이제곱 싸이파이 패키지
from scipy.stats import chi2

In [7]:
np.random.seed(111)
# 카이제곱 활용
sample_std = 2
n = 10 #소표본
data = np.random.normal(loc = 30, scale = sample_std, size = n)
data


array([27.73232333, 30.76863839, 32.99310755, 29.28923541, 28.42493292,
       29.08112218, 29.88166246, 29.291653  , 28.52895389, 27.63212023])

In [8]:
# a모분산 검정을 위한 카이제곱 검정 (모분산을 모르는 경우)
sample_variance = np.var(data, ddof = 1)
sample_variance

2.5270164019138934

In [10]:
chi2_stat = (n-1)*sample_variance/4
p_value = chi2.sf(chi2_stat, df = n-1) # 양측 검정

print("chi2통계:",chi2_stat)
print("pvalue:",p_value)

chi2통계: 5.68578690430626
pvalue: 0.7709074026232983


### 문제 4. 대표본 chi-square 검정 (모분산을 알고 있는 경우)
- 한 제약 회사는 생산하는 약의 순도가 일정해야 한다고 주장합니다. 무작위로 50개의 샘플을 분석한 결과, 표준편차가 0.5%로 나타났습니다. 이 데이터가 모집단의 표준편차 0.4%와 유의미하게 다른지 검정하세요. 유의수준은 0.05로 설정하세요.

- 진도 나가지 않은 부분. 과제 전용 추가

In [11]:
# 데이터 생성
np.random.seed(4)
sample_std = 0.5
n = 50
data = np.random.normal(loc=0, scale=sample_std, size=n)

# 모분산 검정을 위한 chi-square 검정 (모분산이 알려져 있음)
sigma0_squared = 0.4**2
sample_variance = np.var(data, ddof=1)
chi2_stat = (n - 1) * sample_variance / sigma0_squared
p_value = chi2.sf(chi2_stat, df=n - 1) * 2  # 양측 검정

# 결과 출력
print(f"Chi2-Statistic: {chi2_stat}, p-Value: {p_value}")


Chi2-Statistic: 75.89559618366587, p-Value: 0.016367468473539003


# 두 모집단의 비교

## 추정과 검정
- 추정: 차이는 얼마인가?
  - 지지율 비교, 평균 통화시간 비교 등
- 검정: 차이가 있는가?
  - 지렁이와 굼뱅이의 차이가 있는가? 비오는날의 지렁이와 건조한날의 지렁이는 다른 지렁이인가?

- 두 표뵨의 이해의 기초
  - 추정량 -> 점추정, 구간추정(신뢰한계)의 이야기
  - 별도 ppt 제공